pkg/crd/generator/generator.go

```diff
@@ -0,0 +1,177 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package generator
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/ghodss/yaml"
+	extensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/gengo/args"
+	crdutil "sigs.k8s.io/controller-tools/pkg/crd/util"
+	"sigs.k8s.io/controller-tools/pkg/internal/codegen"
+	"sigs.k8s.io/controller-tools/pkg/internal/codegen/parse"
+	"sigs.k8s.io/controller-tools/pkg/util"
+)
+
+// Generator search API resource definition files,
+// and generate CRD install files accordingly.
+type Generator struct {
+	RootPath          string
+	Domain            string
+	Namespace         string
+	SkipMapValidation bool
+}
+
+// ValidateAndInitFields validate and init generator fields.
+func (c *Generator) ValidateAndInitFields() error {
+	var err error
+
+	if len(c.RootPath) == 0 {
+		// Take current path as root path if not specified.
+		c.RootPath, err = os.Getwd()
+		if err != nil {
+			return err
+		}
+	}
+
+	// Validate root path is under go src path
+	if !crdutil.IsUnderGoSrcPath(c.RootPath) {
+		return fmt.Errorf("command must be run from path under $GOPATH/src/<package>")
+	}
+
+	// If Domain is not explicitly specified,
+	// try to search for PROJECT file as a basis.
+	if len(c.Domain) == 0 {
+		for {
+			if crdutil.PathHasProjectFile(c.RootPath) {
+				break
+			}
+
+			if crdutil.IsGoSrcPath(c.RootPath) {
+				return fmt.Errorf("failed to find working directory that contains %s", "PROJECT")
+			}
+
+			c.RootPath = path.Dir(c.RootPath)
+		}
+
+		c.Domain = crdutil.GetDomainFromProject(c.RootPath)
+	}
+
+	// Validate apis directory exists under working path
+	apisPath := path.Join(c.RootPath, "pkg/apis")
+	if _, err := os.Stat(apisPath); err != nil {
+		return fmt.Errorf("error validating apis path %s: %v", apisPath, err)
+	}
+
+	return nil
+}
+
+// Do manages CRD generation.
+func (c *Generator) Do() error {
+	arguments := args.Default()
+	b, err := arguments.NewBuilder()
+	if err != nil {
+		return fmt.Errorf("failed making a parser: %v", err)
+	}
+
+	// Switch working directory to root path.
+	if err := os.Chdir(c.RootPath); err != nil {
```

Comment:
wondering why is this required ? This can lead to awkward experience where caller will end up in a different directory if invoked from path other than root-path ?

Reply:
Nope, it just change the working directory of the internal command runtime. User will not experience a directory change.

Tried to remove it and exec `AddDirRecursive` with absolute path path. but saw error like this:

```
ERROR: logging before flag.Parse: W0710 09:28:49.598102   40574 parse.go:224] Ignoring directory /workspace/src/example.com/pkg/apis: unable to import "/workspace/src/example.com/pkg/apis": import "/workspace/src/example.com/pkg/apis": cannot import absolute path
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x8bb54c]

goroutine 1 [running]:
sigs.k8s.io/controller-tools/vendor/k8s.io/gengo/parser.(*Builder).AddDirRecursive(0xc4200d8640, 0xc4202c4180, 0x23, 0xc4202c4180, 0x23)
	/workspace/src/sigs.k8s.io/controller-tools/vendor/k8s.io/gengo/parser/parse.go:229 +0x16c
sigs.k8s.io/controller-tools/pkg/crd/generator.(*Generator).Do(0xc4202544c0, 0x0, 0x0)
	/workspace/src/sigs.k8s.io/controller-tools/pkg/crd/generator/generator.go:102 +0x1a4
sigs.k8s.io/controller-tools/cmd/crd/cmd.glob..func1(0xe26ba0, 0xe4bfc0, 0x0, 0x0)
	/workspace/src/sigs.k8s.io/controller-tools/cmd/crd/cmd/generate.go:49 +0xe2
sigs.k8s.io/controller-tools/vendor/github.com/spf13/cobra.(*Command).execute(0xe26ba0, 0xe4bfc0, 0x0, 0x0, 0xe26ba0, 0xe4bfc0)
	/workspace/src/sigs.k8s.io/controller-tools/vendor/github.com/spf13/cobra/command.go:766 +0x2c1
sigs.k8s.io/controller-tools/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xe26e00, 0xa9b6d1, 0x46, 0x31)
	/workspace/src/sigs.k8s.io/controller-tools/vendor/github.com/spf13/cobra/command.go:852 
sigs.k8s.io/controller-tools/vendor/github.com/spf13/cobra.(*Command).Execute(0xe26e00, 0xc42007c058, 0x0)
	/workspace/src/sigs.k8s.io/controller-tools/vendor/github.com/spf13/cobra/command.go:800 +0x2b
sigs.k8s.io/controller-tools/cmd/crd/cmd.Execute()
	/workspace/src/sigs.k8s.io/controller-tools/cmd/crd/cmd/root.go:40 +0x2d
main.main()
	/workspace/src/sigs.k8s.io/controller-tools/cmd/crd/main.go:22 +0x20
```

I guess `AddDirRecursive` maybe have certain limitation. As I'm sure `Chdir` has no effect on end users, so did not dig it deep.

Reply:
Ack.